src/lib/alloc.c

```diff
@@ -457,6 +457,19 @@ static void free_block(void *ptr)
 	bool heap_is_full;
 
 	heap = get_heap_from_ptr(ptr);
+
+	/* for scenarios like multi-core, the caller might ask for freeing buffer
+	 * with pointer to uncached buffer while the allocation is from cached
+	 * buffer, try to handle the freeing of this kind of buffer.
+	 */
+	if (!heap && is_uncached(ptr)) {
+		tr_info(&mem_tr, "free_block(): uncached buffer %p, try freeing from its cached address",
+			ptr);
+
+		ptr = uncache_to_cache(ptr);
+		heap = get_heap_from_ptr(ptr);
+	}
```

Comment:
There is no need to get_heap_from_ptr twice since the allocator is always returning cached values.
```c
bool heap_is_full;

/* allocator uses cached values internally */
ptr = uncache_to_cache(ptr);
heap = get_heap_from_ptr(ptr);	

if (!heap) {
		tr_err(&amp;mem_tr, "free_block(): invalid heap = %p, cpu = %d",
		       ptr, cpu_get_id());
}
```

Reply:
> There is no need to get_heap_from_ptr twice since the allocator is always returning cached values.

That's not true, when allocating with SOF_MEM_ZONE_RUNTIME_SHARED, we will get uncached values.

> 
> ```c
> bool heap_is_full;
> 
> /* allocator uses cached values internally */
> ptr = uncache_to_cache(ptr);
> heap = get_heap_from_ptr(ptr);	
> 
> if (!heap) {
> 		tr_err(&amp;mem_tr, "free_block(): invalid heap = %p, cpu = %d",
> 		       ptr, cpu_get_id());
> }
> ```

Reply:
ok, so SOF_MEM_ZONE_RUNTIME_SHARED should be in the comments. i.e.
```c
/* allocator uses cached values internally except for  SOF_MEM_ZONE_RUNTIME_SHARED */
heap = get_heap_from_ptr(ptr);
if (heap)
    free();

/* now try SOF_MEM_ZONE_RUNTIME_SHARED region */
ptr = cache_to_uncache(ptr);
heap = get_heap_from_ptr(ptr);
if (heap)
    free();
```
This is easy to understand compared to the previous version.

Reply:
> ok, so SOF_MEM_ZONE_RUNTIME_SHARED should be in the comments. i.e.
> 
> ```c
> /* allocator uses cached values internally except for  SOF_MEM_ZONE_RUNTIME_SHARED */
> heap = get_heap_from_ptr(ptr);
> if (heap)
>     free();
> 
> /* now try SOF_MEM_ZONE_RUNTIME_SHARED region */
> ptr = cache_to_uncache(ptr);

should be uncache_to_cache().

> heap = get_heap_from_ptr(ptr);

if the previous free() is success we don't need this one more try.

> if (heap)
>     free();
> ```
> 
> This is easy to understand compared to the previous version.

Sure, let me try to make it more understandable, the SOF_MEM_ZONE_RUNTIME_SHARED is in uncache range only on cAVS platforms, so which region is in cached range is really platform specific so I would not make the assumption that only the _ZONE_RUNTIME_SHARED can be in uncached region.  And, let me do the 2nd try only it failed at the first, so let me change it like this:

```c
        /* caller uses the direct address got from the allocator, free it directly */
	heap = get_heap_from_ptr(ptr);

        /* some caller uses uncached address while allocated from cached address, e.g. 
         * SOF_MEM_ZONE_RUNTIME_SHARED region on cAVS platform, one more
         * try to free the momory from the corresponding cached address.
         * /
	if (!heap && is_uncached(ptr)) {
		ptr = uncache_to_cache(ptr);
		heap = get_heap_from_ptr(ptr);
	}

	if (!heap) {
		tr_err(&mem_tr, "free_block(): invalid heap = %p, cpu = %d",
		       ptr, cpu_get_id());
		return;
	}

```

Reply:
ok, the version in the comment above is good.